Interfaces/Base.lproj/Preferences.xib

```diff
@@ -1021,8 +1018,48 @@ Warning: If Sequel Ace and the server do not have at least one cipher suite in c
                         <color key="backgroundColor" name="controlColor" catalog="System" colorSpace="catalog"/>
                     </textFieldCell>
                 </textField>
+                <textField horizontalHuggingPriority="251" verticalHuggingPriority="750" fixedFrame="YES" translatesAutoresizingMaskIntoConstraints="NO" id="Rn3-Bt-QGE">
+                    <rect key="frame" x="142" y="248" width="78" height="16"/>
+                    <autoresizingMask key="autoresizingMask" flexibleMaxX="YES" flexibleMinY="YES"/>
+                    <textFieldCell key="cell" lineBreakMode="clipping" title="SSH Config:" id="Ziz-Vf-7cz">
+                        <font key="font" metaFont="system"/>
+                        <color key="textColor" name="labelColor" catalog="System" colorSpace="catalog"/>
+                        <color key="backgroundColor" name="textBackgroundColor" catalog="System" colorSpace="catalog"/>
+                    </textFieldCell>
+                </textField>
+                <textField horizontalHuggingPriority="251" verticalHuggingPriority="750" fixedFrame="YES" translatesAutoresizingMaskIntoConstraints="NO" id="on9-UJ-WZI">
+                    <rect key="frame" x="226" y="248" width="120" height="16"/>
+                    <autoresizingMask key="autoresizingMask" flexibleMaxX="YES" flexibleMinY="YES"/>
+                    <textFieldCell key="cell" lineBreakMode="clipping" title="/path/to/ssh/config" id="9Yv-Uy-f5M">
+                        <font key="font" metaFont="system"/>
+                        <color key="textColor" name="labelColor" catalog="System" colorSpace="catalog"/>
+                        <color key="backgroundColor" name="textBackgroundColor" catalog="System" colorSpace="catalog"/>
+                    </textFieldCell>
+                    <connections>
+                        <binding destination="117" name="value" keyPath="values.SSHConfigFile" id="wqF-7J-HeC">
+                            <dictionary key="options">
+                                <string key="NSNullPlaceholder">System default</string>
```

Comment:
Perhaps we may want to say "Sequel Ace default" as we aren't using the system ssh config by default!

Reply:
Additionally, after selecting a config file, closing preferences, then quitting and re-opening Sequel Ace, the config file picker once again appears as if no file has been selected. I would look at Connection interface controller to see how we show the name of the previously selected file.
<img width="580" alt="Screen Shot 2020-07-01 at 08 25 55" src="https://user-images.githubusercontent.com/10710367/86243614-afe76780-bb74-11ea-9749-62b65906d20f.png">

Reply:
Ooooo idea for the grant additional files thing - could implement a view just like SSL Cipher Suites that lists all files the app has been granted by security scoped bookmarks. Can have an add button to open a file selector to add additional items to the bookmarks (and thus the list) and can delete them from the list to remove them from our bookmarks! That way people can see exactly what files the app is accessing too (openness fosters good privacy protections).

Reply:
Your idea sounds great! Should we then integrate a way to display the chosen config in this list view or do you think it's better to display it above the list?

The separate option would be more readable, I think, but introduces a kind of redundancy. Maybe if we mark the ssh config with a star or visualize it in another way, the user might be able to quick switch between different bookmarks/configs.

Reply:
> Your idea sounds great! Should we then integrate a way to display the chosen config in this list view or do you think it's better to display it above the list?
> 
> 
> 
> The separate option would be more readable, I think, but introduces a kind of redundancy. Maybe if we mark the ssh config with a star or visualize it in another way, the user might be able to quick switch between different bookmarks/configs.

Oooo I kinda like the idea of choosing a different ssh config by starring it or something. Interesting idea. However, I could see us having more types of files in the future. For instance, Sequel Ace also defines a custom known hosts file although some users may wish to use the system known hosts file or another one entirely. What if either we left it as a separate selector for ssh config (but cleared it if you deleted the respective bookmark), or made the files lost have some of drop down for each row where you could choose what to use the file as (ex. Ssh config file or known hosts file). We'd have to play some games to make sure you could only select one config file at a time.

Perhaps for the simplest implementation we could just show the list of granted files (doesn't even have to be able to remove things from the list for initial implementation, but just to see what files you already granted) and identify the ssh config file with a star or something in the list. Keep the selector for ssh config file as separate for now, adding a selector to grant files.

However, we may need to handle the known hosts file just like the ssh config file to actually fit many use cases. I had forgotten about this complication! This could be just another simple file selector setup just like the ssh config file however!

Reply:
Hmm, then I think that the star idea might get confusing. Which files exactly do you want to list in there? All bookmarked files? If we implement it this way, we should consider to implement a new settings pane for it, as then also ssh/ssl keys could be revoked.

However, I will then try to implement the known hosts file as will in this pull request. For the selection, I will try to use a drop down menu for each of them, which contains all granted files already. One of the items will then be an "Other file..." entry, which will pop up with the currently implemented file picker. With this solution we avoid the confusion and caveats with selecting multiple config files from a list, but maintain the quick switch feature.

How do you think about this?

Reply:
I agree the star may get confusing. A new settings pane is probably the right move! For this to keep it small, we could make the list of files only show granted files separate from the config/known hosts and key files?

Reply:
That’s what I suppose. One kind of „security“ or files tab, that displays all the granted files (including the bookmarks e.g. for the SSL certificates etc.). In the network tab then two drop down menus for choosing the ssh config and known hosts file.

I‘ll give it a try tomorrow

Reply:
Hmmm thinking about this a little bit I wonder if we'd want these settings to actually be configurable per-connection. I could see someone wanting to use different known config/hosts files for different connections without having to go to settings to change.

Reply:
For the known host file, wouldn't this weaken the security purpose of this file? E.g. when a user has two profiles for the same host. If he configures different host files, he might not be able to detect, whether the host has the same key or not. It would also enable a fast way to ignore known host errors. I'm not sure if it is still a sufficient security for our use-case.

I agree with you for the config file. It would enable a cool way to separate configs for different purposes. However, I'm not sure if the separation is normally performed in the ssh config file with "Include" directives. This would enable the separation for lots of other applications as well.

Reply:
Yes and no on known hosts weakening - if the user chooses to overwrite the known hosts file for a connection, as long as both connections point at the same known hosts file it shouldn't weaken security. I would imagine most users would be choosing to use their system known hosts file which they would likely share on purpose. However, allowing custom known hosts file selection allows special use cases where the user may need to specify a known hosts file that is handled/modified by some external process without weakening the security of changing the functionality of other connections in the app

Reply:
I've now implemented the drop down menu and the file preference pane(just not the toolbar icon, since I don't know where to pick it from 😄).

![Screen Shot 2020-07-02 at 18 48 37](https://user-images.githubusercontent.com/4348959/86388190-a5bc8a80-bc94-11ea-8450-405420d5f423.png)

![Screen Shot 2020-07-02 at 18 46 59](https://user-images.githubusercontent.com/4348959/86388075-6b52ed80-bc94-11ea-8090-b5d9d9bb4f10.png)

It was a bit tricky to sync the bookmarks between the preferences and the connection editor. I don't know, if this way is a good implementation or if you know some other way. Both features seem to work quite nice for me. Is this the way you had in mind?

If it's ok for you, I'd like to split the know-hosts file apart from this pull request, as this one only was supposed to implement the ssh config? When we implement this feature in this pull request, it might look confusing later in the commit history, if the commits get squashed.

Reply:
It sounds good to keep known hosts file as a separate task! I could come in soon after this is merged and easily leverage it to do known hosts and custom default id rsa file!

Reply:
>It was a bit tricky to sync the bookmarks between the preferences and the connection editor.

What was the problem? The bookmarks are stored in NSUserDefaults, so should be available from both screens. Ages ago you could force a write to file but nowadays thats a NO OP. I'll have a look at the latest commits.

Reply:
Works great. Are you suuuure you're a beginner? 😃

Reply:
>  Oh now I see what you mean. Could add KVO to the prefs object and only reload if changed, but that wouldn't save much time.

This exactly was the problem. When I edited the files in the preferences, then saved a public key file in the connection editor, it will override the bookmarks for the config file. My first intuition was, that there should be some kind of mechanism for properties (like `ObservableObject` in SwiftUI), but haven't found something. Since I do not know the patterns for this in obj-c, I just implemented a reload every time, they go saved or used.

Thanks a lot for the tip with the KVO! I will give it a try. Maybe it won't save a lot time, but seems to be the "right" and cleaner way.

> Works great. Are you suuuure you're a beginner? 😃

At least in macOS and objective c 😄 I'm just trying to transfer my knowledge of other languages(like C/C++, Go, PHP, Python) into this. Thanks! Actually it's the first time I'm writing obj-c. Never thought, this pr will get through the review 😅

Reply:
The KVO is working great! Thanks a lot!